### ISSUE TITLE
Graph pattern arrangements on POST/REACT pair: Clarification

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#424]
In GitLab by @facoelho on Apr 13, 2023, 13:27

Hi. 
We (@vasco.m.campos @ruben.m.queiros @carlos.m.pereira )in the PT pilot setup are experiencing something I feel is not natural to the KE and would like to confirm if I am "reasoning" correctly about it. 

![Screenshot_2023-04-13_at_14.18.48](https://gitlab.inesctec.pt/interconnect/knowledge-engine/uploads/1b052a1ae6ac11db7dae7be99723985b/Screenshot_2023-04-13_at_14.18.48.png)

Consider this picture, that represents the ongoing setup for one of our services.
We have 2 KBs. KB1 has one Post KI (KI-A) while KB2 has 2 react KIs, KI-C and KI-D. 
All KIs have the same argument GP, but only KI-A and KI-C have the same result GP. 

From what we read on the documentation, the matching occurs (besides the KIs pair: post/react or ask/answer and the communicative act) if **both** the argument and result GPs match. This means that if Post KI-A is picked up by a react KI, in this case only KI-C should be activated to reply. Is this correct? 

Currently what we are witnessing is that both react KIs are being activated.

Our primary intention is to have two possible answers and that we could handle it by issuing the post with the 'type' determined by the result graph pattern.
Moreover on an isolated test, i.e., there are not handlers in parallel for both react KIs, I am able to activate both react KIs. I was expecting that only KI-C would be activated. 

Is this expected / are we modelling this incorrectly from the get go ?
Best.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 14, 2023, 07:36

Thanks for raising this @facoelho 

It is indeed surprising that the `KI-D` knowledge interaction is also triggered here. With the *matcher* I would not have expected that to be the case because both the argument and result patterns should fully match. With the *reasoner* enabled, it __is__ expected that `KI-D` is triggered, since its result pattern fully covers `KI-A`'s result pattern and the reasoner should be able to see that. (Notice how the first 4 triple patterns in `KI-D`'s result pattern are equivalent to the other result patterns.)

With the *matcher* situation, I wonder what happens when the result pattern of `KI-D` is completely different. Would `KI-A` then receive an empty binding? I'll try to test this in a new unit test to make sure we don't unintentionally do stuff like that.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 14, 2023, 11:54

created branch [`410-graph-pattern-arrangements-on-post-react-pair-clarification`](/interconnect/knowledge-engine/-/compare/master...410-graph-pattern-arrangements-on-post-react-pair-clarification) to address this issue

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 14, 2023, 12:14

I've reproduced a similar situation in a unit test, and observed the following:

- indeed both REACT KI's are triggered
- both REACT KI's are included in the result binding set of the POST KI, but the binding that was produced by the KI with the bigger graph pattern is stripped completely empty (so it is an empty binding).

Now, to decide on the intended behaviour:

- ~~I think it makes sense that the REACT KI with the bigger result pattern is triggered, since it 'subscribed' to the event in the argument pattern. The REACTing KB internally might write this to a database or show it in a GUI for example. And even if it produces result bindings (that are potentially ignored; see below) it should still have received the event, in my view.~~
- ~~but it doesn't make sense that this results in an *empty* binding in the POSTing KI. The POSTing KI should be able to expect all result bindings that it gets to be *full* (so all variables have a value).~~

I've updated my new unit test to assert the above (REACTing KI with bigger result pattern should receive the event, but the resulting bindings should not be in the POST result) and it currently fails because the empty binding __is__ included in the result.

@barry.nouwt.tno.nl do you agree with my conclusions?

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Apr 14, 2023, 14:48

Hi @han.kruiger.tno.nl, I disagree with your conclusion :smile_cat:. Especially when using the matcher, I think it is easier if interactions only occur when both the `argument` and `result` graph pattern match. So, this is, I think, in line with the expectation that @facoelho expresses.

When the reasoner is involved, it becomes a bit more complex. My first intuition is that at least a single triple pattern from the argument graph pattern of a REACT KI should be used, otherwise that particular KI is not triggered. In my opinion, however, this is out of scope for this issue.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 14, 2023, 15:45

That makes sense to keep it simpler. I will strike out my previous conclusions.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 17, 2023, 07:52

So, new conclusions about the things in scope for this issue:

- We agree that, when using the matcher, the REACT KI should only be triggered when *both* the argument and result pattern match exactly with a POST KI.
- The current implementation of the matcher also triggers REACT KI's that have only a matching argument pattern, and this should be fixed to not be the case.

I think this is indeed in line with what @facoelho expects. Can you confirm that?

**Comment by Sophietje** [comment]
In GitLab by @facoelho on Apr 17, 2023, 08:39

@han.kruiger.tno.nl indeed our expectation and design of the KIs considered full matches one both argument and response KI.
How should we proceed given the time?

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 17, 2023, 08:57

If this is important for the pilot I can increase the priority and hopefully get it working before the next release.

**Comment by Sophietje** [comment]
In GitLab by @facoelho on Apr 17, 2023, 09:07

When is the release expected?

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 17, 2023, 10:57

April 28th is the current goal.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 17, 2023, 15:16

We're a bit afraid that partners in the InterConnect project may rely on the REACT KI being triggered in the case when only the argument pattern matches.

So it may be worth considering to not put this in a release for InterConnect. Is that workable for you?

**Comment by Sophietje** [comment]
In GitLab by @facoelho on Apr 17, 2023, 15:36

OK.
But @han.kruiger.tno.nl, to be completely sure, this means that:
- we can have 2 REACT KIs with the same argument GP and completely different result GPs (i.e., one result GP is not a sub-set of the other)?
- We need to ensure that the argument GP of both REACT KIs is not a match ? (the match here is a string comparison, correct e.g., 'KIC-argumentGP.equals(KID-argumentGP) )

which applies?

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 18, 2023, 07:16

The current situation is that:

- You can have 2 REACT KIs with the same argument graph pattern, but a different result graph pattern. And they can both be triggered by the same POST KI as long as the argument graph pattern matches.
- If you only want 1 of the REACT KIs to be triggered, you'll need to change one of the REACT KIs' argument pattern so that it doesn't match with the POST KI anymore.

With *matching*, I mean the usual graph pattern matching. __So not a string comparison.__ The same triples patterns should be used in both graph patterns. Variable names can be different, and order of the triple patterns can be different, but then they can still be matching.

E.g., this graph pattern:

```
?var1 <http://example.org/pred1> ?var2 .
?var1 <http://example.org/pred2> ?var3 .
```

matches with this one:

```
?varA <http://example.org/pred2> ?varB .
?varA <http://example.org/pred1> ?varC .
```

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 28, 2023, 07:47

@facoelho What is your final call on this? Shall we include the fix in the 1.2.0 release? @barry.nouwt.tno.nl and I think that the best way forward is to include it. If users depend on the (buggy) behaviour then they can always keep using the older version.

**Comment by Sophietje** [comment]
In GitLab by @facoelho on Apr 28, 2023, 08:06

Let go forward and include it. 
As Barry stated in his email, then the decision is to update the KE or not.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Apr 28, 2023, 11:40

mentioned in commit c52e619bd8d533526a53b93b052440deaadc278e
